File: src/app/browser/main/tab/page/content/text.cpp
#include "text.hpp"
#include "text/gemini.hpp"
#include "text/plain.hpp"

using namespace app::browser::main::tab::page::content;

Text::Text(
    const Type & TYPE,
    const Glib::ustring & REQUEST,
    const Glib::ustring & SOURCE
) {
    switch (TYPE)
    {
        case GEMINI:

            set_child(
                * Gtk::make_managed<text::Gemini>(
                    REQUEST,
                    SOURCE,
                    title
                )
            );

        break;

        case PLAIN:

            set_child(
                * Gtk::make_managed<text::Plain>(
                    SOURCE
                )
            );

        break;

        default:

            throw _("Invalid text type enum"); // @TODO
    }
}

// Getters
Glib::ustring Text::get_title()
{
    return title;
}

File: src/browser/main/tab/page/content/text.rs
mod gemini;

use gemini::Gemini;

use gtk::{
    glib::{GString, Uri},
    ScrolledWindow,
};

pub struct Meta {
    title: Option<GString>,
}

pub struct Text {
    meta: Meta,
    widget: ScrolledWindow,
}

impl Text {
    // Construct
    pub fn gemini(gemtext: &str, base: &Uri) -> Self {
        // Init components
        let gemini = Gemini::new(gemtext, base);

        // Init meta
        let meta = Meta {
            title: gemini.reader_title().clone(),
        };

        // Init widget
        let widget = ScrolledWindow::builder().build();

        widget.set_child(Some(gemini.widget()));

        // Result
        Self { meta, widget }
    }

    // Getters
    pub fn meta_title(&self) -> &Option<GString> {
        &self.meta.title
    }

    pub fn widget(&self) -> &ScrolledWindow {
        &self.widget
    }
}
